### DISCUSSION TITLE
How can I disable padding?

### DISCUSSION THREAD

**Discussion by UnsignedInt8** [krzyzanowskim/CryptoSwift#963] [General]
I want to disable data padding, and I passed nil to padding, but no effect. Please tell me how to disable data padding, thank you.

**Comment by krzyzanowskim** [comment]
use nil with code from branch 'develop'

**Comment by UnsignedInt8** [comment]
Thank you, I'm going to try it

**Comment by UnsignedInt8** [comment]
Hi, by using develop branch to disable data padding, it works. But the encrypted data can't be decrypted by Node.js.

The first two bytes can be decrypted, but the last encrypted data can't be decrypted.

<img width="963" alt="screen shot 2015-12-07 at 09 22 28" src="https://cloud.githubusercontent.com/assets/15274326/11617042/991e78dc-9cc4-11e5-9cf4-1881b68bb38c.png">

```
let (cipher, iv) = Crypto.createCipher(self.cipherAlgorithm, password: self.password)

let meta: [UInt8] = [0x05, UInt8(drand48() * 0xff)]
let randomBytes = AES.randomIV(Int(meta[1]))

var bytes = [UInt8](count: data.length, repeatedValue: 0)
data.getBytes(&bytes) // Read data from socket

let encrypted = NSMutableData()
encrypted.appendData(NSData(bytes: iv)) // IV, no encrypted
encrypted.appendData(NSData(bytes: try! cipher.encrypt(meta, padding: nil)))  // Ok, can be decrypted by Node.js
encrypted.appendData(NSData(bytes: randomBytes)) // Random Padding Bytes, no encrypted

// The data can't be decrypted by Node.js
>> encrypted.appendData(NSData(bytes: try! cipher.encrypt(bytes, padding: nil)))  // Error, can't be decrypted by Node.js

proxy.writeData(encrypted, withTimeout: -1, tag: 0)
```

Here is protocol:

```
------------------Request---------------------

+------+------+------+----------+------------+
| IV   | TYPE | PLEN | RPADDING | SOCKETDATA |
+------+------+------+----------+------------+
| 8-16 | 1    | 1    | 0-255    | VARIABLE   |
+------+------+------+----------+------------+
| RAW  | ENCRYPTED   | RAW      | ENCRYPTED  |
+------+-------------+----------+------------+

```

Is my code writing wrong?

**Comment by dickwu** [comment]
![image](https://cloud.githubusercontent.com/assets/4208695/11617208/80edec28-9cc6-11e5-96ac-b2a2e8139bc2.png)
not padding : 
let jiares = try! jia.encrypt(byte!)

**Comment by UnsignedInt8** [comment]
@dickwu The problem is: the encrypted data can't be decrypted by Node.js.  The first two bytes can be decrypted, but the last data CAN'T be decrypted.

**Comment by dickwu** [comment]
http://8gwifi.org/CipherFunctions.jsp
use this website to debug

**Comment by UnsignedInt8** [comment]
I tested my code, but I don't know where wrong:

IV from client

<img width="242" alt="0 iv generated from client" src="https://cloud.githubusercontent.com/assets/15274326/11617887/3b62ebfa-9ccf-11e5-8ab1-d6c98f89888d.png">

Server received IV

<img width="369" alt="0 iv by server" src="https://cloud.githubusercontent.com/assets/15274326/11617891/45ab7fa0-9ccf-11e5-9312-0d869ed246c0.png">

No problem with IV.

---

Sent encrypted data from client

<img width="698" alt="1 encrypted by client" src="https://cloud.githubusercontent.com/assets/15274326/11617893/5b48e370-9ccf-11e5-832b-26e5e916bb90.png">

Server received data

<img width="413" alt="2 received from client" src="https://cloud.githubusercontent.com/assets/15274326/11617901/7220ba64-9ccf-11e5-9502-56d3da500b53.png">

No problem with encrypted data

---

Server decrypted data

<img width="419" alt="3 decrypted by node js" src="https://cloud.githubusercontent.com/assets/15274326/11617914/98daf7f0-9ccf-11e5-8372-c27b9ee42841.png">

Client decrypted data

<img width="780" alt="4 decrypted by client raw data" src="https://cloud.githubusercontent.com/assets/15274326/11617916/a8029e40-9ccf-11e5-802f-6ca8802a6f7c.png">

The decrypted data not same, why? But the first two can be decrypted by server! Cipher algorithm and cipher key are same.

First two bytes sent from client:

<img width="233" alt="5 first two bytes" src="https://cloud.githubusercontent.com/assets/15274326/11617976/320b2e7c-9cd0-11e5-87a5-1fa652562df1.png">

First two bytes decrypted by server:

<img width="350" alt="5 frist two bytes decrypted by server" src="https://cloud.githubusercontent.com/assets/15274326/11617986/47d03de2-9cd0-11e5-818c-e4b32854b537.png">

**Comment by dickwu** [comment]
![image](https://cloud.githubusercontent.com/assets/4208695/11618115/af177b86-9cd1-11e5-94e0-db8c2c3c9e25.png)
if you have qq please add my qq :"374026384"

**Comment by UnsignedInt8** [comment]
<img width="472" alt="screen shot 2015-12-07 at 11 03 08" src="https://cloud.githubusercontent.com/assets/15274326/11618163/20c7ba5c-9cd2-11e5-9e05-5641ff0113ab.png">

I used CFB block-mode.

**Comment by dickwu** [comment]
I need the iv func to debug

**Comment by UnsignedInt8** [comment]
IV function ?

I just use AES.randomIV(16) to generate IV. 

<img width="898" alt="screen shot 2015-12-07 at 11 14 37" src="https://cloud.githubusercontent.com/assets/15274326/11618287/bfa7fce4-9cd3-11e5-810d-4fe132d3d63a.png">

@dickwu I don't use QQ, can we use email to communicate?

**Comment by dickwu** [comment]
ok my email is dickwu@vip.qq.com

**Comment by dickwu** [comment]
![image](https://cloud.githubusercontent.com/assets/4208695/11618791/3721263c-9cda-11e5-91b3-a481e7a4ef38.png)

**Comment by UnsignedInt8** [comment]
@krzyzanowskim Hi, author. Do you know why Node.js can't decrypt it?

**Comment by krzyzanowskim** [comment]
@SunshinyNeko I see a lot of problem how you use Node and CryptoSwift and convert types. From what I see you encrypt encrypt invalid data with this code.

**Comment by UnsignedInt8** [comment]
Encrypt invalid data? Could you tell me how to do it rightly?

**Comment by UnsignedInt8** [comment]
@krzyzanowskim I know where wrong, thank you, you are right. :)
